cep47/src/data.rs

```diff
@@ -0,0 +1,300 @@
+    }
+
+    pub fn remove(&self, key: &TokenId) {
+        self.dict.remove::<Key>(key);
+    }
+}
+
+pub struct Metadata {
+    dict: Dict,
+}
+
+impl Metadata {
+    pub fn instance() -> Metadata {
+        Metadata {
+            dict: Dict::at(METADATA_DICT),
+        }
+    }
+
+    pub fn get(&self, key: &TokenId) -> Option<Meta> {
+        self.dict.get(key)
+    }
+
+    pub fn set(&self, key: &TokenId, value: Meta) {
+        self.dict.set(key, value);
+    }
+
+    pub fn remove(&self, key: &TokenId) {
+        self.dict.remove::<Meta>(key);
+    }
+}
+
+pub fn name() -> String {
+    get_key("name").unwrap_or_revert()
+}
+
+pub fn symbol() -> String {
+    get_key("symbol").unwrap_or_revert()
+}
+
+pub fn meta() -> Meta {
+    get_key("meta").unwrap_or_revert()
+}
+
+pub fn total_supply() -> U256 {
+    get_key("total_supply").unwrap_or_revert()
+}
+
+pub fn update_total_supply(total_supply: U256) {
+    set_key("total_supply", total_supply);
+}
+
+pub fn is_paused() -> bool {
+    get_key("paused").unwrap()
+}
+
+pub fn pause() {
+    set_key("paused", true);
+}
+
+pub fn unpause() {
+    set_key("paused", false);
+}
+
+pub fn get_nonce() -> u32 {
+    get_key("nonce").unwrap_or_default()
+}
+
+pub fn set_nonce(nonce: u32) {
+    set_key("nonce", nonce);
+}
+
+pub fn contract_package_hash() -> ContractPackageHash {
+    get_key("contract_package_hash").unwrap_or_revert()
+}
+
+pub fn initial_named_keys(
+    package_hash: ContractPackageHash,
+    token_name: &str,
+    token_symbol: &str,
+    token_meta: Meta,
+    paused: bool,
+) -> NamedKeys {
+    let mut named_keys = NamedKeys::new();
+    named_keys.insert("name".to_string(), storage::new_uref(token_name).into());
+    named_keys.insert("symbol".to_string(), storage::new_uref(token_symbol).into());
+    named_keys.insert("meta".to_string(), storage::new_uref(token_meta).into());
+    named_keys.insert(
+        "total_supply".to_string(),
+        storage::new_uref(U256::zero()).into(),
+    );
+    named_keys.insert("paused".to_string(), storage::new_uref(paused).into());
+    named_keys.insert(
+        "contract_package_hash".to_string(),
+        storage::new_uref(package_hash).into(),
+    );
+
+    // Add empty dictionaries.
+    add_empty_dict(&mut named_keys, BALANCES_DICT);
+    add_empty_dict(&mut named_keys, OWNED_TOKENS_DICT);
+    add_empty_dict(&mut named_keys, TOKEN_OWNERS_DICT);
+    add_empty_dict(&mut named_keys, METADATA_DICT);
+
+    named_keys
+}
+
+fn add_empty_dict(named_keys: &mut NamedKeys, name: &str) {
+    let dict = new_dictionary(name).unwrap_or_revert();
+    runtime::remove_key(name);
+    named_keys.insert(name.to_string(), dict.into());
+}
+
+fn key_to_str(key: &Key) -> String {
+    match key {
+        Key::Account(account) => account.to_string(),
+        Key::Hash(package) => hex::encode(package),
+        _ => runtime::revert(ApiError::UnexpectedKeyVariant),
+    }
+}
+
+pub fn emit(event: &CEP47Event) {
```

Comment:
Where and how do we test this?

Reply:
`cep47-js` has tests for that. I still want to experiment with different ways of throwing events. Then I'll proper tests.